From c7913e142a09a1f1e935b580305ba93968501804 Mon Sep 17 00:00:00 2001
From: Nirbheek Chauhan <nirbheek@centricular.com>
Date: Fri, 16 Aug 2019 13:56:39 +0530
Subject: [PATCH] gio: Disable win32 API and modules when unavailable on WinRT

The modules listed below are useless on a UWP app:

* gwin32filemonitor
* gwin32volumemonitor
* gwinhttpvfs

The following APIs are also useless:

* gwin32mount
* gwin32registrykey

Don't build them at all so that apps get errors at build time instead
of runtime.

TODO: the following need to be ported to new WinRT APIs:

* gwin32networkmonitor (giomodule)
* gwin32appinfo.c (currently stubbed out)
---
 gio/gappinfo.h  |  5 +++++
 gio/giomodule.c | 32 +++++++++++++++++---------------
 gio/meson.build | 33 ++++++++++++++++++---------------
 3 files changed, 40 insertions(+), 30 deletions(-)

diff --git a/gio/gappinfo.h b/gio/gappinfo.h
index 84162f0..5ca8bf5 100644
--- a/gio/gappinfo.h
+++ b/gio/gappinfo.h
@@ -282,6 +282,11 @@ GLIB_AVAILABLE_IN_2_50
 gboolean  g_app_info_launch_default_for_uri_finish (GAsyncResult         *result,
                                                     GError              **error);
 
+#ifdef G_WINAPI_ONLY_APP
+#define g_app_info_get_default_for_type(x,y) NULL
+#define g_app_info_get_default_for_uri_scheme(x) NULL
+#define g_app_info_get_all() NULL
+#endif
 
 /**
  * GAppLaunchContext:
diff --git a/gio/giomodule.c b/gio/giomodule.c
index 0860499..1f8042c 100644
--- a/gio/giomodule.c
+++ b/gio/giomodule.c
@@ -1066,14 +1066,10 @@ _g_io_module_get_default (const gchar         *extension_point,
 
 extern GType g_inotify_file_monitor_get_type (void);
 extern GType g_kqueue_file_monitor_get_type (void);
-extern GType g_win32_file_monitor_get_type (void);
 
 extern GType _g_unix_volume_monitor_get_type (void);
 extern GType _g_local_vfs_get_type (void);
 
-extern GType _g_win32_volume_monitor_get_type (void);
-extern GType _g_winhttp_vfs_get_type (void);
-
 extern GType _g_dummy_proxy_resolver_get_type (void);
 extern GType _g_dummy_tls_backend_get_type (void);
 extern GType g_network_monitor_base_get_type (void);
@@ -1103,6 +1099,11 @@ extern GType g_cocoa_notification_backend_get_type (void);
 #ifdef G_PLATFORM_WIN32
 extern GType g_win32_notification_backend_get_type (void);
 
+#ifndef G_WINAPI_ONLY_APP
+extern GType g_win32_file_monitor_get_type (void);
+extern GType _g_win32_volume_monitor_get_type (void);
+extern GType _g_winhttp_vfs_get_type (void);
+
 #include <windows.h>
 extern GType _g_win32_network_monitor_get_type (void);
 
@@ -1162,6 +1163,7 @@ giomodule_init_ctor (void)
 #error Your platform/compiler is missing constructor support
 #endif /* GLIB_STATIC_COMPILATION */
 
+/* Currently only used in gdbusprivate.c */
 void *
 _g_io_win32_get_module (void)
 {
@@ -1187,7 +1189,9 @@ _g_io_modules_ensure_extension_points_registered (void)
 #if !GLIB_CHECK_VERSION (3, 0, 0)
       ep = g_io_extension_point_register (G_DESKTOP_APP_INFO_LOOKUP_EXTENSION_POINT_NAME);
       g_io_extension_point_set_required_type (ep, G_TYPE_DESKTOP_APP_INFO_LOOKUP);
-#endif
+#endif /* !G_WINAPI_ONLY_APP */
+
+#endif /* G_PLATFORM_WIN32 */
 #endif
 
       ep = g_io_extension_point_register (G_LOCAL_FILE_MONITOR_EXTENSION_POINT_NAME);
@@ -1252,7 +1256,7 @@ get_gio_module_dir (void)
   module_dir = !is_setuid ? g_strdup (g_getenv ("GIO_MODULE_DIR")) : NULL;
   if (module_dir == NULL)
     {
-#ifdef G_OS_WIN32
+#if defined(G_OS_WIN32) && !defined(G_WINAPI_ONLY_APP)
       gchar *install_dir;
 
       install_dir = g_win32_get_package_installation_directory_of_module (gio_dll);
@@ -1333,10 +1337,15 @@ _g_io_modules_ensure_loaded (void)
       g_type_ensure (g_kqueue_file_monitor_get_type ());
 #endif
 #ifdef G_OS_WIN32
+      g_type_ensure (g_win32_notification_backend_get_type ());
+#ifndef G_WINAPI_ONLY_APP
+      g_type_ensure (_g_win32_network_monitor_get_type ());
       g_type_ensure (_g_win32_volume_monitor_get_type ());
-      g_type_ensure (g_win32_file_monitor_get_type ());
+      g_type_ensure (_g_winhttp_vfs_get_type ());
       g_type_ensure (g_registry_backend_get_type ());
-#endif
+      g_type_ensure (g_win32_file_monitor_get_type ());
+#endif /* !G_WINAPI_ONLY_APP */
+#endif /* G_OS_WIN32 */
 #ifdef HAVE_COCOA
       g_type_ensure (g_nextstep_settings_backend_get_type ());
       g_type_ensure (g_osx_app_info_get_type ());
@@ -1355,10 +1364,6 @@ _g_io_modules_ensure_loaded (void)
 #endif
 #if MAC_OS_X_VERSION_MIN_REQUIRED >= 1090
       g_type_ensure (g_cocoa_notification_backend_get_type ());
-#endif
-#ifdef G_OS_WIN32
-      g_type_ensure (g_win32_notification_backend_get_type ());
-      g_type_ensure (_g_winhttp_vfs_get_type ());
       g_type_ensure (g_memory_monitor_win32_get_type ());
 #endif
       g_type_ensure (_g_local_vfs_get_type ());
@@ -1374,9 +1379,6 @@ _g_io_modules_ensure_loaded (void)
       g_type_ensure (_g_network_monitor_netlink_get_type ());
       g_type_ensure (_g_network_monitor_nm_get_type ());
 #endif
-#ifdef G_OS_WIN32
-      g_type_ensure (_g_win32_network_monitor_get_type ());
-#endif
 
       g_once_init_leave (&loaded_dirs, TRUE);
     }
diff --git a/gio/meson.build b/gio/meson.build
index 14f0d82..56a9289 100644
--- a/gio/meson.build
+++ b/gio/meson.build
@@ -272,7 +272,7 @@ settings_sources = files(
   'gsettings.c',
 )
 
-if host_system == 'windows'
+if host_system == 'windows' and not glib_conf.has('G_WINAPI_ONLY_APP')
   settings_sources += files('gregistrysettingsbackend.c')
 endif
 
@@ -428,30 +428,34 @@ if host_system != 'windows'
     )
   endif
 else
-  appinfo_sources += files('gwin32appinfo.c')
   contenttype_sources += files('gcontenttype-win32.c')
-  platform_deps += [cc.find_library('shlwapi'),
-                    cc.find_library('dnsapi'),
-                    iphlpapi_dep,
-                    winsock2]
   platform_deps += uwp_gio_deps
 
   win32_sources += files(
-    'gmemorymonitorwin32.c',
-    'gwin32registrykey.c',
-    'gwin32mount.c',
-    'gwin32volumemonitor.c',
+    'gmemorymonitorwin32.c', # Maybe this neeeds to be moved into WINAPI_ONLY_APP section as well? (tpm) 
     'gwin32inputstream.c',
     'gwin32outputstream.c',
     'gwin32file-sync-stream.c',
     'gwin32packageparser.c',
-    'gwin32networkmonitor.c',
-    'gwin32networkmonitor.h',
     'gwin32notificationbackend.c',
     'gwin32sid.c',
     'gwin32sid.h',
   )
 
+  if not glib_conf.has('G_WINAPI_ONLY_APP')
+    platform_deps += [cc.find_library('shlwapi'),
+                      cc.find_library('dnsapi'),
+                      iphlpapi_dep,
+                      winsock2]
+    appinfo_sources += files('gwin32appinfo.c')
+    win32_sources += files(
+      'gwin32mount.c',
+      'gwin32registrykey.c',
+      'gwin32volumemonitor.c',
+      'gwin32networkmonitor.c'
+    )
+  endif
+
   gio_win_rc = configure_file(
     input: 'gio.rc.in',
     output: 'gio.rc',
@@ -551,6 +555,7 @@ gio_sources = files(
   'gsimpleiostream.c',
   'gsimplepermission.c',
   'gsimpleproxyresolver.c',
+  'gthreadedresolver.c',
   'gsocket.c',
   'gsocketaddress.c',
   'gsocketaddressenumerator.c',
@@ -570,8 +575,6 @@ gio_sources = files(
   'gtcpwrapperconnection.c',
   'gthemedicon.c',
   'gthreadedsocketservice.c',
-  'gthreadedresolver.c',
-  'gthreadedresolver.h',
   'gtlsbackend.c',
   'gtlscertificate.c',
   'gtlsclientconnection.c',
@@ -782,7 +785,7 @@ if have_func_kqueue and have_func_kevent
   internal_deps += [ kqueue_lib ]
 endif
 
-if host_system == 'windows'
+if host_system == 'windows' and not glib_conf.has('G_WINAPI_ONLY_APP')
   subdir('win32')
   internal_deps += [ giowin32_lib ]
 endif
-- 
2.38.1

